pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexService.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+import com.google.common.collect.Sets;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.Set;
+
+
+/**
+ * This is the entry point of the Index SPI.
+ *
+ * Ideally, if we used some kind of injection system, this class should be injected into a Pinot context all classes can
+ * receive when they are built. Given that Pinot doesn't have that, we have to relay on static fields.
+ *
+ * By default, this class will be initialized by reading all ServiceLoader SPI services that implement
+ * {@link IndexPlugin}, adding all the {@link IndexType} that can be found in that way.
+ *
+ * In case we want to change the instance to be used at runtime, the method {@link #setInstance(IndexService)} can be
+ * called.
+ */
+public class IndexService {
```

Comment:
Is there a reason why an instance of this class need to exist? Could _allIndexes be initialised statically and `getOrThrow(String indexId)` be a static method?

Reply:
We have been discussing about this in https://github.com/apache/pinot/pull/10192 and also in the PEP ([link](https://docs.google.com/document/d/1L_GJ2tAgZVxI_Ssa59--Mso3mK_HeAS7s7G9261-sOE/edit?disco=AAAAsjZUDEA)).

To have singletons or statics is something we don't like, but AFAIK is the only thing we can do right now. My suggestion is to have a context we can inject in places where we need these contextual information. Alternatively we could use ThreadLocal. The idea here is to have the logic in object methods instead that in class methods in order to be closer to the ideal world without statics.

Meanwhile, by having the instance, we can write the code in a way that it can be tested without static components. I mean, we can consume this class with `IndexService.getInstance().getAllIndexes()` or we can receive the instance as a construction argument and it an attribute to later do `_indexService.getAllIndexes()`. If we do the later, we can easily create test scenarios with different index type sets. Otherwise to create a test that do that we need to modify the static variable, which means that it is more difficult (if even possible) to run tests in parallel.

I have to admit that the code I wrote follows the first pattern instead of the later, but it is easy to change.